File: c++/find-minimum-in-rotated-sorted-array-ii.cpp
//https://oj.leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/

class Solution {
private:
    int findMinHelper(vector<int> &num, int start, int end) {
        if (start > end) {
            return INT_MAX;
        }
        if (start == end || num[start] < num[end]) {
            return num[start];
        }
        if (start + 1 == end) {
            return min(num[start], num[end]);
        }

        int mid = start + (end - start) / 2;

        if (num[mid] == num[start] && num[mid] == num[end]) {
            return min(findMinHelper(num, start, mid-1), findMinHelper(num, mid+1, end));
        }

        if (num[mid] > num[start] || num[mid] > num[end]) {  // num[mid] located in the left-part
            return findMinHelper(num, mid+1, end);
        }
        else {
            return findMinHelper(num, start, mid);
        }
    }

public:
    int findMin(vector<int> &num) {
        // the worst time complexity is O(n),
        // "worst" situations, e.g., [2,2,2,...,2,2,1,2,...,2,2,2]
        return findMinHelper(num, 0, num.size()-1);
    }
};

File: python/find-minimum-in-rotated-sorted-array-ii.py
class Solution:
    # @param num, a list of integer
    # @return an integer
    def findMin(self, num):
        return self._findMinHelper(num, 0, len(num)-1)

    def _findMinHelper(self, num, start, end):
        if start == end:
            return num[start]
        if start + 1 == end:
            return min(num[start], num[end])
        if num[start] < num[end]:
            return num[start]

        mid = start + (end - start) / 2
        if num[mid] > num[end]:
            return self._findMinHelper(num, mid+1, end)
        elif num[mid] < num[start]:
            return self._findMinHelper(num, start, mid)

        return min(self._findMinHelper(num, mid+1, end), self._findMinHelper(num, start, mid-1))
